src/AstTranslator.cpp

```diff
@@ -303,40 +303,42 @@ std::unique_ptr<RamCondition> AstTranslator::translateConstraint(
+                for (size_t height = 1; height < auxiliaryArity; height++) {
+                    values.push_back(translator.translateValue(args[arity + height], index));
+                }
             }
-            if (arity > 0) {
-                return std::make_unique<RamNegation>(std::make_unique<RamExistenceCheck>(
-                        translator.translateRelation(atom), std::move(values)));
-            } else {
-                return std::make_unique<RamEmptinessCheck>(translator.translateRelation(atom));
-            }
+            return std::make_unique<RamNegation>(std::make_unique<RamProvenanceExistenceCheck>(
+                    translator.translateRelation(atom), std::move(values)));
         }
 
-        /** for provenance negation */
-        std::unique_ptr<RamCondition> visitProvenanceNegation(const AstProvenanceNegation& neg) override {
+        /** for negations */
+        std::unique_ptr<RamCondition> visitNegation(const AstNegation& neg) override {
             const auto* atom = neg.getAtom();
             size_t auxiliaryArity = translator.getEvaluationArity(atom);
-            assert(auxiliaryArity < atom->getArity() && "auxiliary arity out of bounds");
+            assert(auxiliaryArity <= atom->getArity() && "auxiliary arity out of bounds");
             size_t arity = atom->getArity() - auxiliaryArity;
             std::vector<std::unique_ptr<RamExpression>> values;
 
-            auto args = atom->getArguments();
-            for (size_t i = 0; i < arity; i++) {
-                values.push_back(translator.translateValue(args[i], index));
-            }
-            // we don't care about the provenance columns when doing the existence check
-            if (Global::config().has("provenance")) {
-                // undefined value for rule number
-                values.push_back(std::make_unique<RamUndefValue>());
-                // add the height annotation for provenanceNotExists
-                for (size_t h = 0; h + 1 < auxiliaryArity; h++) {
-                    values.push_back(translator.translateValue(args[arity + h + 1], index));
+            if (arity == 0) {
+                // for a nullary, negation is a simple emptiness check
+                return std::make_unique<RamEmptinessCheck>(translator.translateRelation(atom));
+            } else {
+                // else, we construct the atom and create a negation
+                auto args = atom->getArguments();
+                for (size_t i = 0; i < arity; i++) {
+                    values.push_back(translator.translateValue(args[i], index));
                 }
+                for (size_t i = 0; i < auxiliaryArity; i++) {
+                    values.push_back(std::make_unique<RamUndefValue>());
+                }
+                return std::make_unique<RamNegation>(std::make_unique<RamExistenceCheck>(
+                        translator.translateRelation(atom), std::move(values)));
             }
```

Comment:
This is better, but I was actually trying to suggest a simpler version with an early exit if arity == 0, like:
```suggestion
            }
            // else, we construct the atom and create a negation
            std::vector<std::unique_ptr<RamExpression>> values;
            auto args = atom->getArguments();
            for (size_t i = 0; i < arity; i++) {
                values.push_back(translator.translateValue(args[i], index));
            }
            for (size_t i = 0; i < auxiliaryArity; i++) {
                values.push_back(std::make_unique<RamUndefValue>());
            }
            return std::make_unique<RamNegation>(std::make_unique<RamExistenceCheck>(
                    translator.translateRelation(atom), std::move(values)));
            }
```